sdk/go/pulumi/provider.go

```diff
@@ -0,0 +1,792 @@
+	port, done, err := rpcutil.Serve(0, nil, []func(*grpc.Server) error{
+		func(srv *grpc.Server) error {
+			pulumirpc.RegisterResourceProviderServer(srv, prov)
+			return nil
+		},
+	}, nil)
+	if err != nil {
+		return errors.Errorf("fatal: %v", err)
+	}
+
+	// The resource provider protocol requires that we now write out the port we have chosen to listen on.
+	fmt.Printf("%d\n", port)
+
+	// Finally, wait for the server to stop serving.
+	if err := <-done; err != nil {
+		return errors.Errorf("fatal: %v", err)
+	}
+
+	return nil
+}
+
+type providerServer struct {
+	version       string
+	args          ProviderArgs
+	engineConn    *grpc.ClientConn
+	keepSecrets   bool
+	keepResources bool
+}
+
+// ProviderArgs allows specifying implementations of the Provider interface for orchestrating resource create, read,
+// update, and delete operations.
+type ProviderArgs struct {
+	// Schema is the schema for the provider in JSON.
+	Schema []byte
+
+	// Check validates that the given property bag is valid for a resource of the given type and returns the inputs
+	// that should be passed to successive calls to Diff, Create, or Update for this resource.
+	Check func(ctx *ProviderContext, urn resource.URN, olds,
+		news resource.PropertyMap) (resource.PropertyMap, []plugin.CheckFailure, error)
+
+	// Diff checks what impacts a hypothetical update will have on the resource's properties.
+	Diff func(ctx *ProviderContext, urn resource.URN, id resource.ID, olds resource.PropertyMap,
+		news resource.PropertyMap, ignoreChanges []string) (plugin.DiffResult, error)
+
+	// Create allocates a new instance of the provided resource and returns its unique resource.ID.
+	Create func(ctx *ProviderContext, urn resource.URN, inputs resource.PropertyMap, timeout float64,
+		preview bool) (resource.ID, resource.PropertyMap, error)
+
+	// Read the current live state associated with a resource.  Enough state must be include in the inputs to uniquely
+	// identify the resource; this is typically just the resource ID, but may also include some properties.  If the
+	// resource is missing (for instance, because it has been deleted), the resulting property map will be nil.
+	Read func(ctx *ProviderContext, urn resource.URN, id resource.ID,
+		props resource.PropertyMap) (resource.ID, resource.PropertyMap, error)
+
+	// Update updates an existing resource with new values.
+	Update func(ctx *ProviderContext, urn resource.URN, id resource.ID,
+		olds resource.PropertyMap, news resource.PropertyMap, timeout float64,
+		ignoreChanges []string, preview bool) (resource.PropertyMap, error)
+
+	// Delete tears down an existing resource.
+	Delete func(ctx *ProviderContext, urn resource.URN, id resource.ID, props resource.PropertyMap,
+		timeout float64) error
+
+	// Construct creates a new component resource.
+	Construct func(ctx *Context, typ, name string, inputs *ConstructInputs,
+		options ResourceOption) (ConstructResult, error)
```

Comment:
I didn't want to force authors to have to implement an entire provider interface just to support `Construct`, so I've made it so you pass these as fields on this struct.

Reply:
No longer applicable.